underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java

```diff
@@ -0,0 +1,406 @@
+    if (mUploadId == null) {
+      initMultiPartUpload();
+    }
+    if (b == null || len == 0) {
+      return;
+    }
+    if (mFile == null) {
+      initNewFile();
+    }
+    if (mOffset + len < mPartitionSize) {
+      mLocalOutputStream.write(b, off, len);
+      mOffset += len;
+    } else {
+      int firstLen = (int) (mPartitionSize - mOffset);
+      mLocalOutputStream.write(b, off, firstLen);
+      mOffset += firstLen;
+      uploadPart();
+      write(b, off + firstLen, len - firstLen);
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    // We try to minimize the time use to close()
+    // because Fuse release() method which calls close() is async.
+    // In flush(), we upload the current writing file if it is bigger than 5 MB,
+    // and wait for all current upload to complete.
+    mLocalOutputStream.flush();
+    if (mOffset > UPLOAD_THRESHOLD) {
+      uploadPart();
+    }
+    waitForAllPartsUpload();
+    mFutureTags = new ArrayList<>();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    try {
+      if (mFile != null) {
+        mLocalOutputStream.close();
+        int partNumber = mPartNumber.getAndIncrement();
+        final UploadPartRequest uploadRequest = new UploadPartRequest()
+            .withBucketName(mBucketName)
+            .withKey(mKey)
+            .withUploadId(mUploadId)
+            .withPartNumber(partNumber)
+            .withFile(mFile)
+            .withPartSize(mFile.length());
+        uploadRequest.setLastPart(true);
+        execUpload(uploadRequest, mFile, partNumber);
+      }
+
+      waitForAllPartsUpload();
+      completeMultiPartUpload();
+    } catch (Exception e) {
+      LOG.error("Failed to upload {}: {}", mKey, e.toString());
+      throw new IOException(e);
+    }
+    // Set the closed flag, close can be retried until mFile.delete is called successfully
+    mClosed = true;
+  }
+
+  /**
+   * Initializes multipart upload.
+   */
+  private void initMultiPartUpload() throws IOException {
+    // Generate the object metadata by setting server side encryption, md5 checksum,
+    // and encoding as octet stream since no assumptions are made about the file type
+    ObjectMetadata meta = new ObjectMetadata();
+    if (SSE_ENABLED) {
+      meta.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+    }
+    if (mHash != null) {
+      meta.setContentMD5(new String(Base64.encode(mHash.digest())));
+    }
+    meta.setContentType(Mimetypes.MIMETYPE_OCTET_STREAM);
+
+    AmazonClientException lastException;
+    InitiateMultipartUploadRequest initRequest =
+        new InitiateMultipartUploadRequest(mBucketName, mKey).withObjectMetadata(meta);
+    do {
+      try {
+        mUploadId = mClient.initiateMultipartUpload(initRequest).getUploadId();
+        return;
+      } catch (AmazonClientException e) {
+        lastException = e;
+      }
+    } while (mRetryPolicy.attempt());
```

Comment:
I want the whole upload request can retry for 5 times instead of each operation could retry for 5 times. If the whole request occurs errors more than 5 times, we better to cancel it and clean up.